### DISCUSSION TITLE
How to cast VariantType (datatype) when batch writing to UA Server

### DISCUSSION THREAD

**Discussion by ryangu18** [FreeOpcUa/python-opcua#1450] [Q&A] [answered]
I read data in batches from a source UA server and write to a destination UA server. I used code like this.

`
results = clientSour.uaclient.get_attributes(nodeids_Sour, ua.AttributeIds.Value)  #read
results1 = clientDes.uaclient.set_attributes(nodeids_Des, results, ua.AttributeIds.Value)  #write
`

results value is

![Snipaste_2022-03-29_16-51-32](https://user-images.githubusercontent.com/99937589/160573269-d91b871f-3720-4700-bdf1-14d0e87ee42b.png)

I have a problem, in the source UA server these data have various types, double, Float, int32 ect, but in the destination UA server, all data types are double.
![Snipaste_2022-03-29_16-53-48](https://user-images.githubusercontent.com/99937589/160573583-ce7b01f0-1ca2-4fd2-8d87-4a17630ddb8a.png)

Data writing failed due to inconsistent data types.
I cannot modify the data type of the destination UA server. Can I cast the different data type of the source UA server to double, then write to destination UA server?

Thanks in advance

**Comment by AndreasHeine** [comment, marked as answer]
you could read the DataType of the tags in Dest UA Server up front after connected and Map it to the NodeIds of the Source Server (a simple dict "SrcNodeId": "DstVariantType") and then loop one time over the batch and create the list of DataValues with the new VariantTypes and with the old Values and Timestamps?

**Reply by ryangu18** [reply]
Thank you very much for your suggestion, it is indeed a good approach.
However, I think this may reduce the efficiency of the program. this program is to write more than 1000 tags per second. Before batch reading and writing, I tried to read and write tags one by one. The speed is very slow and the CPU usage is very high. When I changed to batch read and write, the speed became very fast.

If I need to replace the VariantTypes of the source UA server with the VariantTypes of the destination UA server one by one every second, it seems to return to the tag-by-tag operation, right?

So, can the function ignore VariantTypes in batch writes, which seems to be a better approach. But I don't have the documentation, I'm not sure if this idea will work.

I know you are an expert on this topic and would be very happy to hear from you.

**Reply by AndreasHeine** [reply]
> If I need to replace the VariantTypes of the source UA server with the VariantTypes of the destination UA server one by one every second, it seems to return to the tag-by-tag operation, right?

well, its more like 
1. batchread get result list
2. loop over result list (one time) create new list
3. batchwrite new list

> So, can the function ignore VariantTypes in batch writes, which seems to be a better approach. But I don't have the documentation, I'm not sure if this idea will work.

the important thing to know is that the dst server decides what he accepts (he owns the information model)! if he expect a double you need to send a double and if not, you will have a BAD statuscode in response from the server so there is nothing you can do on clientside!

#### hint:
For batch-read/write you should check the servers operational limits after connection (MaxNodesPerWrite and MaxNodesPerRead) and only request that amount in one request!

**Reply by ryangu18** [reply]
thank you very much.

**Reply by AndreasHeine** [reply]
> But the speed does drop, because of the loop, I have 100 data, and the delay reaches 4s. Do you think my code can still be optimized?

4s for 100 DataPoints ??? sounds like something else is wrong! you can use time.time() before the loop and after the loop to check how long it takes! you need to figure out the time for batchread (request/response), for the loop itself and for batchwrite (request/response)!

`ua.uatypes.DataValue(float(result.Value.Value), result.StatusCode, result.SourceTimestamp))`
thats not what i meant in the first place ^^ 
float() is unnecessary in my opinion you need to create the variant according to the dest server datatype!
if you dont provide a variant to the datavalue it trys to guess the type but this is not always a correct guess so its better to do it like this:


```python
ua.DataValue(ua.Variant(result.Value.Value, ua.VariantType.Float), result.StatusCode, result.SourceTimestamp))

# ua.VariantType.Float must come from the dst server nodes which you need to read after connecting to the dst server and prepare a dict which map the src nodeid to the dst datatype
```
